## Review

Reviewed with two independent passes plus verification of every claim against a running repro. The direction is right, but the first version had two regressions that were worse than the hang it fixed. Both are now fixed in a7ee564a with tests that fail against the previous code.

### Fixed during review

**1. `concurrentLimiter.Close` could panic the process.** When `Close` returned on timeout, its `wg.Wait()` helper stayed parked. The abandoned request eventually calls `wg.Done()`, waking that waiter — and any `wg.Add(1)` landing in between panics with `sync: WaitGroup is reused before previous Wait has returned`. `ScheduleRequest` had no closed flag, and `SyncWithPeer` is explicitly documented as safe to call after close (`headsync.go:114`), so the `Add` is reachable. Reproduced:

```
panic: sync: WaitGroup is reused before previous Wait has returned
	.../keyvalue/limiter.go:63
created by (*concurrentLimiter).Close
```

Fixed by rejecting requests once closed.

**2. `ocache.Close` left roughly half the healthy entries unclosed.** The shared deadline was passed to `remove`, which calls `waitLoad` first. For a loaded entry both `<-ctx.Done()` and `<-e.load` are ready, and Go's `select` picks at random — so once the budget expired, every remaining entry had a coin-flip chance of returning `ctx.Err()` before `value.Close()` was ever called, and was left in `c.data` with `c.closed` already true, where nothing can ever close it. Measured 91–116 of 200 entries abandoned. In this repo that is `net/pool`'s caches: one stuck peer burns the budget, then healthy peers keep their conns and `acceptLoop` goroutines after `pool.Close()`. Nondeterministic per run, so it would have surfaced as an intermittent field bug.

Fixed by spending the deadline only on the wait-for-another-closer, with loads on an uncancellable ctx (`Close` already cancels them just above). The comment claiming the deadline bounds the whole pass was wrong and has been corrected — `value.Close()` takes no ctx, so total close time still scales with cache size.

**3. `deleteLoop.Close` burned the full 10s when `Run` never ran** — the same close-before-run case the PR guards for `subscribeClient`, `actionPool` and `streamPool`. Guarded now.

### Confirmed sound

- **webtransport `defer cancel()`** does not kill the returned session: `HandshakeOutbound` builds `cctx` from `context.Background()` (`secureservice.go:171`), and webtransport-go re-parents the session with `context.WithoutCancel`. Same pattern yamux already uses.
- **GO-7332 / ba235868 invariants hold.** The ctx escape in `setClosing` only drops a waiter — it never assigns `e.close` or claims ownership, and `remove` returns before `value.Close()`/`setClosed`. No double close of a channel or a value.
- **`peer.gc`** deferred close is correctly ordered (LIFO puts it after `Unlock`), and conns leave `active`/`inactive` under the lock, so no double close.
- No API breakage: everything changed is unexported, and the `OCache` interface and all `Close(ctx)` signatures are unchanged.

### Open, deliberately not addressed here

1. **The per-component budgets are additive and can still blow the watchdog.** Each space now gets 10s (deletionmanager) + 10s (keyvalue) of hard timer, and `commonspace/space.go:250` passes `context.Background()`, so every `ctx.Done()` arm added here is dead in production — only the timers apply. Three spaces closing serially reaches `app.StopDeadline` (60s) on its own. The real fix is a single space-level budget at `space.go:250` with the per-component timers as backstops, and constants meaningfully below it. That changes shutdown semantics for all ~15 space components, so it wants its own change.
2. **`syncAcl.Close` still takes an unbounded `s.Lock()`.** Cancelling `broadcastCtx` unblocks it only if the client's `peerManager` honours ctx — and `peermanager` is client-implemented, the same category as `treemanager`, which this PR deliberately does not trust.
3. **`settingsObject.addContent` now widens an existing torn-write window.** `AddContentWithValidator` advances the in-memory tree (`objecttree.go:294`) before persisting (`:299`), so a cancelled client ctx can leave a head that was never written. Any `AddAll` failure already does this, so it is pre-existing rather than new, but ctx cancellation is a new trigger on a client-facing path.
4. **Timeouts are hardcoded where this repo's convention is config.** quic and webtransport both expose `CloseTimeoutSec`; `ocache` already has `WithTTL`/`WithGCPeriod` options. All three new constants are exactly `app.StopWarningAfter` (10s), so any one firing guarantees the "components close in progress" warning.
5. **yamux `StreamCloseTimeout` reuses `WriteTimeoutSec`.** Semantically correct — it RSTs only the stream, not the session — but it is a different quantity, and lowering the write timeout would silently start resetting merely-slow streams. Deserves its own config field.
6. **Untested behaviour**: the syncacl broadcast ctx (no `Close` test exists in that package at all), the deleter `ctx.Err()` bail-outs, the settings ctx plumbing, `peer.gc`'s post-unlock close, and both transport timeouts.
7. **Downstream risk this strategy assumes**: any-store's `ConnManager.Close` closes read conns without waiting for checked-out ones, so a goroutine still executing a statement hits a closed sqlite handle. Before this PR no component goroutine outlived its `Close`; now they can. Not fixable inside any-sync, but the embedder should know.

### Verdict

Mergeable once the reviewer agrees on (1) — the additive budgets are the difference between "bounded" and "bounded per component but not overall". Everything else above is follow-up.

--- a/app/ocache/ocache.go
+++ b/app/ocache/ocache.go
@@ -221,10 +221,16 @@ func (c *oCache) closeAndDelete(e *entry) {
 }
 
 func (c *oCache) remove(ctx context.Context, e *entry) (ok bool, err error) {
-	if _, err = e.waitLoad(ctx, e.id); err != nil {
+	return c.removeCtx(ctx, ctx, e)
+}
+
+// loadCtx bounds waiting for an in-flight load, closingCtx bounds waiting for
+// another closer to release the entry
+func (c *oCache) removeCtx(loadCtx, closingCtx context.Context, e *entry) (ok bool, err error) {
+	if _, err = e.waitLoad(loadCtx, e.id); err != nil {
 		return false, err
 	}
-	_, curState, err := e.setClosing(ctx, true)
+	_, curState, err := e.setClosing(closingCtx, true)
 	if err != nil {
 		return false, err
 	}
@@ -405,13 +411,14 @@ func (c *oCache) Close() (err error) {
 		toClose = append(toClose, e)
 	}
 	c.mu.Unlock()
-	// one deadline for the whole close: an entry being closed by the gc can be
-	// stuck in TryClose on an unresponsive peer, and waiting per entry would let
-	// the total grow with the cache size
-	ctx, cancel := context.WithTimeout(context.Background(), c.closeTimeout)
+	// one deadline for the whole pass, spent only on entries another closer holds:
+	// that closer can be a gc stuck in TryClose on an unresponsive peer. Loads are
+	// already cancelled above, and value.Close takes no ctx, so uncontended entries
+	// still close normally once the deadline has passed.
+	closingCtx, cancel := context.WithTimeout(context.Background(), c.closeTimeout)
 	defer cancel()
 	for _, e := range toClose {
-		if _, err := c.remove(ctx, e); err != nil && err != ErrNotExists {
+		if _, err := c.removeCtx(context.Background(), closingCtx, e); err != nil && err != ErrNotExists {
 			c.log.With("object_id", e.id).Warnf("cache close: object close error: %v", err)
 		}
 	}

--- a/app/ocache/ocache_test.go
+++ b/app/ocache/ocache_test.go
@@ -778,3 +778,35 @@ func TestOCache_CloseBoundedByOtherCloser(t *testing.T) {
 		require.Fail(t, "Close blocked behind an entry owned by another closer")
 	}
 }
+
+// The close deadline is spent only on entries another closer holds: healthy
+// entries must still be closed after it has expired.
+func TestOCache_CloseClosesHealthyEntriesAfterDeadline(t *testing.T) {
+	var objects []*testObject
+	c := New(func(ctx context.Context, id string) (Object, error) {
+		o := NewTestObject(id, true, nil)
+		objects = append(objects, o)
+		return o, nil
+	})
+	oc := c.(*oCache)
+	oc.closeTimeout = time.Millisecond
+	for i := 0; i < 50; i++ {
+		_, err := c.Get(ctx, fmt.Sprint(i))
+		require.NoError(t, err)
+	}
+	// one entry is held by another closer and will eat the whole deadline
+	oc.mu.Lock()
+	held := oc.data["0"]
+	oc.mu.Unlock()
+	_, _, err := held.setClosing(ctx, false)
+	require.NoError(t, err)
+
+	require.NoError(t, c.Close())
+	var notClosed []string
+	for _, o := range objects {
+		if o.name != "0" && !o.closeCalled {
+			notClosed = append(notClosed, o.name)
+		}
+	}
+	require.Empty(t, notClosed, "entries nobody else holds must still be closed after the deadline")
+}

--- a/commonspace/deletionmanager/deleteloop.go
+++ b/commonspace/deletionmanager/deleteloop.go
@@ -2,6 +2,7 @@ package deletionmanager
 
 import (
 	"context"
+	"sync/atomic"
 	"time"
 
 	"go.uber.org/zap"
@@ -19,6 +20,7 @@ type deleteLoop struct {
 	deleteFunc   func(ctx context.Context)
 	loopDone     chan struct{}
 	closeTimeout time.Duration
+	running      atomic.Bool
 }
 
 func newDeleteLoop(deleteFunc func(ctx context.Context)) *deleteLoop {
@@ -34,6 +36,7 @@ func newDeleteLoop(deleteFunc func(ctx context.Context)) *deleteLoop {
 }
 
 func (dl *deleteLoop) Run() {
+	dl.running.Store(true)
 	go dl.loop()
 }
 
@@ -67,6 +70,11 @@ func (dl *deleteLoop) notify() {
 // here forever wedges the whole app.Close.
 func (dl *deleteLoop) Close(ctx context.Context) {
 	dl.deleteCancel()
+	// loopDone is closed by loop, which only Run starts: app.Start closes
+	// components it never ran
+	if !dl.running.Load() {
+		return
+	}
 	timer := time.NewTimer(dl.closeTimeout)
 	defer timer.Stop()
 	select {

--- a/commonspace/deletionmanager/deleteloop_test.go
+++ b/commonspace/deletionmanager/deleteloop_test.go
@@ -78,3 +78,18 @@ func TestDeleteLoop_CloseHonoursCtx(t *testing.T) {
 	}
 	close(release)
 }
+
+// app.Start closes components it never ran: loopDone is only closed by the loop.
+func TestDeleteLoop_CloseWithoutRun(t *testing.T) {
+	dl := newDeleteLoop(func(ctx context.Context) {})
+	closed := make(chan struct{})
+	go func() {
+		dl.Close(context.Background())
+		close(closed)
+	}()
+	select {
+	case <-closed:
+	case <-time.After(time.Second):
+		require.Fail(t, "Close waited for a loop that was never started")
+	}
+}

--- a/commonspace/object/keyvalue/limiter.go
+++ b/commonspace/object/keyvalue/limiter.go
@@ -14,6 +14,7 @@ type concurrentLimiter struct {
 	mu           sync.Mutex
 	inProgress   map[string]bool
 	wg           sync.WaitGroup
+	closed       bool
 	closeTimeout time.Duration
 }
 
@@ -26,7 +27,10 @@ func newConcurrentLimiter() *concurrentLimiter {
 
 func (cl *concurrentLimiter) ScheduleRequest(ctx context.Context, id string, action func()) bool {
 	cl.mu.Lock()
-	if cl.inProgress[id] {
+	// a bounded Close can return while wg still has waiters parked: an Add after
+	// that panics with "WaitGroup is reused before previous Wait has returned",
+	// and SyncWithPeer stays callable after Close
+	if cl.closed || cl.inProgress[id] {
 		cl.mu.Unlock()
 		return false
 	}
@@ -54,10 +58,14 @@ func (cl *concurrentLimiter) ScheduleRequest(ctx context.Context, id string, act
 	return true
 }
 
-// Close waits for the scheduled requests to finish. The wait is bounded: a request
-// already past the ctx check is doing peer rpc that may not return while nodes are
-// unreachable, and blocking here forever wedges the whole app.Close.
+// Close rejects further requests and waits for the scheduled ones to finish. The
+// wait is bounded: a request already past the ctx check is doing peer rpc that may
+// not return while nodes are unreachable, and blocking here forever wedges the
+// whole app.Close. On timeout the request is abandoned, not stopped.
 func (cl *concurrentLimiter) Close(ctx context.Context) {
+	cl.mu.Lock()
+	cl.closed = true
+	cl.mu.Unlock()
 	done := make(chan struct{})
 	go func() {
 		cl.wg.Wait()

--- a/commonspace/object/keyvalue/limiter_test.go
+++ b/commonspace/object/keyvalue/limiter_test.go
@@ -2,6 +2,7 @@ package keyvalue
 
 import (
 	"context"
+	"sync"
 	"testing"
 	"time"
 
@@ -57,4 +58,31 @@ func TestConcurrentLimiter_CloseHonoursCtx(t *testing.T) {
 		require.Fail(t, "Close ignored the cancelled ctx")
 	}
 	close(release)
+}
+
+// A timed-out Close leaves a waiter parked on the WaitGroup: a later Add would
+// panic with "WaitGroup is reused before previous Wait has returned".
+func TestConcurrentLimiter_NoScheduleAfterClose(t *testing.T) {
+	for i := 0; i < 500; i++ {
+		cl := newConcurrentLimiter()
+		cl.closeTimeout = time.Microsecond
+		release := make(chan struct{})
+		require.True(t, cl.ScheduleRequest(context.Background(), "peer", func() {
+			<-release
+		}))
+		cl.Close(context.Background())
+
+		var start sync.WaitGroup
+		start.Add(2)
+		go func() {
+			start.Done()
+			start.Wait()
+			close(release)
+		}()
+		go func() {
+			start.Done()
+			start.Wait()
+			require.False(t, cl.ScheduleRequest(context.Background(), "peer2", func() {}))
+		}()
+	}
 }

